[clang] [Sema] Preserve nested name specifier prefix in MemberPointerType

HighCommander4 opened the pull request:
Fixes https://github.com/llvm/llvm-project/issues/118198

HighCommander4 committed:
[clang] [Sema] Preserve nested name specifier prefix in MemberPointerType

HighCommander4 commented:
(Marked as Draft since there are test failures I need to investigate.)

mizvekov commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -5347,15 +5347,18 @@ static TypeSourceInfo *GetFullTypeForDeclarator(TypeProcessingState &state,
+          // DependentTemplateSpecialization, it already stores the NNS prefix.
+          // Otherwise, wrap it in an Elaborated type to have a place to store
+          // the NNS prefix.
+          if (!(isa<ElaboratedType>(NNSType) ||
+                isa<DependentNameType>(NNSType) ||
+                isa<DependentTemplateSpecializationType>(NNSType))) {
```
This seems to be testing for conditions which seem impossible, the NNS would be malformed otherwise.

For example, what would it mean for the NNS type to be an ElaboratedType? If the NNS has a prefix, which would take precedence, that prefix or the ElaboratedType's prefix? If there is no prefix, what sense would it convey beyond using the Elaborated prefix as its prefix instead?

For DependentNameType, that would be redundant with how we represent these as Identifier + prefix, so we should prefer that over forming a NNS with such type.

The DTST is an exception here, as we lack a NNS kind for an identifier + list of template arguments, but then such DTST should not store a NNS itself, otherwise again this would be a bug somewhere else, as that would be redundant / confusing with the NNS' own prefix as explained above.

So if we are getting either an ElaboratedType or a DependentNameType here, there seems to be a bug elsewhere.

Regarding the explanation for DependentTemplateSpecialization: even if the type node itself can have a nested name, it should not be present when the node is used as the type for a NNS node itself.

We should handle a DTST here by rebuilding it, adding the NNSPrefix as the DTST's nested name.

HighCommander4 replied:
@mizvekov Thanks for having a look!

I don't have any evidence of us actually getting an ElaboratedType or DependentNameType here, I was just basing this on the comment which mentioned these types.

I revised the patch to make a more targeted change (wrap `RecordType` when there is a qualifier present), which still addresses the original issue I was having in clangd, and (hopefully) passes tests. I would be grateful of your opinion about whether there is a more appropriate condition we should be using.

mizvekov replied:
Yeah I think the existing comment is wrong / misleading.

mizvekov commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -5347,13 +5347,16 @@ static TypeSourceInfo *GetFullTypeForDeclarator(TypeProcessingState &state,
+          // already is an Elaborated, DependentName, or
+          // DependentTemplateSpecialization.
+          if (isa<TemplateSpecializationType>(NNSType) ||
+              (NNSPrefix && isa<RecordType>(NNSType)))
```
```suggestion
              isa<RecordType>(NNSType))
```
We use an ElaboratedType with null prefix to represent a type which was written without qualifiers, so it seems arbitrary to not add the elaboration in this case.

Also, can you at least add a FIXME / TODO explaining how a DependentTemplateSpecializationType needs to be handled here, as I explained in the other thread?

mizvekov replied:
It would also be cool to have an assertion here, that we are handling all cases.

I think the logic here would be something like:
```
if (isa<DependentTemplateSpecializationType>...)
  // Rebuild DependentTemplateSpecializationType, adding the Prefix.
else
  // either the dependent case (TemplateSpecializationType), or the non-dependent one (RecordType).
 // assert(isa<TemplateSpecializationType, RecordType>...)
```

HighCommander4 replied:
> We use an ElaboratedType with null prefix to represent a type which was written without qualifiers, so it seems arbitrary to not add the elaboration in this case.

Wrapping RecordTypes into ElaboratedTypes with a null prefix was causing a bunch of failures on the first draft of the patch.

For example, on this line:

https://github.com/llvm/llvm-project/blob/8cb44859cc31929521c09fc6a8add66d53db44de/clang/test/CXX/conv/conv.mem/p4.cpp#L22

the change has the effect of changing the way the derived type is printed from `test1::Derived` to just `Derived`, I guess because when the diagnostic formatter encounters an `ElaboratedType` it says "I know how the type was printed in the source, I will print it the same way in the diagnostic".

Should we accept changes like this to diagnostic output?

mizvekov replied:
Yes exactly, this test change is correct, because the type was written as 'Derived' and not 'test1::Derived', and the diagnostics are supposed to print the type as written. It would be inconsistent to print the type as written when it has any qualifier, but print it fully qualified when it does not have any.

We started doing this consistently since this patch: https://reviews.llvm.org/D112374

HighCommander4 replied:
> It would also be cool to have an assertion here, that we are handling all cases.
> 
> I think the logic here would be something like:
> 
> ```
> if (isa<DependentTemplateSpecializationType>...)
>   // Rebuild DependentTemplateSpecializationType, adding the Prefix.
> else
>   // either the dependent case (TemplateSpecializationType), or the non-dependent one (RecordType).
>  // assert(isa<TemplateSpecializationType, RecordType>...)
> ```

I tried this, but `SemaCXX/err_init_conversion_failed.cpp` trips the assertion; we get a `TemplateTypeParmType` here.

mizvekov replied:
I see, well this is one case I left out :) The assert is still viable and would be helpful, but it's not critical for merging this.

HighCommander4 replied:
> The assert is still viable and would be helpful

FWIW, I did try adding back the assert to see what other sorts of types we get in this codepath. I found the following:

 * `TemplateTypeParmType` (seen in SemaCXX/err_init_conversion_failed.cpp)
 * `TypedefType` (seen in CXX/drs/cwg3xx.cpp)
 * `DecltypeType` (seen in CXX/expr/expr_prim/expr_prim_general/p8-0x.cpp)
 * `InjectedClassNameType` (seen in CXX/temp/temp_decls/temp_variadic/p5.cpp)
 * `UsingType` (seen in Modules/using-decl.cpp)

It's not clear to me what sort of handling we'd want for these various types.

HighCommander4 commented:
I updated the patch to do the wrapping into `ElaboratedType` even if a prefix isn't present, and updated the various failing tests accordingly.

I haven't added an assert, as explained in the previous comment.

HighCommander4 commented:
Fixed one additional test failure.

HighCommander4 commented:
CI is green. Requesting review.

mizvekov reviewed (approved):
Thanks, LGTM!
